emission/analysis/intake/segmentation/trip_segmentation_methods/dwell_segmentation_dist_filter.py

```diff
@@ -24,112 +19,150 @@
+        This function examines a range of the timeseries data and returns a list of
+        segmentation points (as pairs of start and end points) that define trips.
+        
+        The algorithm computes time differences, distance differences (using the haversine function)
+        and speeds between successive points. In addition, it computes two flags:
+          - 'tracking_restarted': whether the tracking has been restarted (e.g. due to app restart)
+          - 'ongoing_motion': whether there is evidence of motion activity between points.
+        
+        Candidate segmentation points are those where:
+          - The time gap exceeds the time_threshold, AND
+          - One or more of the following conditions hold:
+              * Tracking has restarted,
+              * There is no ongoing motion,
+              * The time gap is extremely large (e.g. > 12 hours), or
+              * The computed speed is below a threshold (indicating little movement)
+          - AND the distance gap is at least the distance_threshold.
+        
+        A placeholder for a row-by-row check is included for huge invalid timestamp offsets,
+        although currently no individual point is invalidated.
+        
+        Finally, the candidate points are used to split the filtered points into trips.
+        
+        :param timeseries: The timeseries database interface
+        :param time_query: The time range query to select relevant entries.
+        :param loc_df: A pandas DataFrame containing filtered location points.
+        :return: A list of tuples. Each tuple contains a start and end point (wrapped as AttrDict)
+                 that demarcate a trip.
         """
-        with ect.Timer() as t_get_filtered_points:
-            self.filtered_points_df = filtered_points_df
-            user_id = self.filtered_points_df["user_id"].iloc[0]
-        esds.store_pipeline_time(
-            user_id,
-            ecwp.PipelineStages.TRIP_SEGMENTATION.name + "/segment_into_trips_dist/get_filtered_points_df",
-            time.time(),
-            t_get_filtered_points.elapsed
+        user_id = loc_df["user_id"].iloc[0]
+        loc_df['valid'] = True
+        loc_df['index'] = loc_df.index
+
+        self.transition_df = transition_df
+        self.motion_df = motion_df
+
+        # --- Compute vectorized differences ---
+        # Calculate the time difference (in seconds) between consecutive points.
+        loc_df['ts_diff'] = loc_df['ts'].diff()
+        # Calculate the distance difference between consecutive points using the haversine function.
+        loc_df['dist_diff'] = haversine_numpy(
+            loc_df['longitude'].shift(1).to_numpy(), loc_df['latitude'].shift(1).to_numpy(),
+            loc_df['longitude'].to_numpy(), loc_df['latitude'].to_numpy()
         )
+        # Compute speed (meters per second) from the distance and time differences.
+        loc_df['speed'] = loc_df['dist_diff'] / loc_df['ts_diff']
```

Comment:
As a future fix, it would be good to replace the existing `add_speed_dist...` function with this vectorized implementation. That should speed up `JUMP_SMOOTHING` as well

Reply:
https://github.com/e-mission/e-mission-docs/issues/1105#issuecomment-2705738660